lib/vrl/compiler/src/vm/machine.rs

```diff
@@ -0,0 +1,541 @@
+    /// set by the ensuing primitive instruction.
+    JumpIfTrue,
+
+    /// If the error field of the VM is not set advances the instruction pointer by the amount
+    /// set by the ensuing primitive instruction.
+    JumpIfNotErr,
+
+    /// Unconditionally advances the instruction pointer by the amount set by the ensuing
+    /// primitive instruction.
+    Jump,
+
+    /// Takes the path indicated by the ensuing primitive and sets this path with the value
+    /// at the top of the stack. The value is not removed from the stack so it can continue
+    /// to be used.
+    SetPath,
+
+    /// Sets either a success or error path. The next primitive is the index to the target for
+    /// the success path, the one after is the error path. After that is a pointer to a
+    /// constant indicating the default value to set to the success path should there be an error.
+    SetPathInfallible,
+
+    /// Takes the ensuing primitive as a pointer to a path. Retrieves this value from the state
+    /// and pushes this onto the stack.
+    GetPath,
+
+    /// Calls the function indicated by the ensuing primitive.
+    Call,
+
+    /// Creates an array. The ensuing primitive indicates the number of elements in the array.
+    /// This amount of values are popped from the stack. The resulting array is then pushed back on
+    /// the stack.
+    CreateArray,
+
+    /// Creates an object. The ensuing primitive indicates the number of elements in the object.
+    /// This amount of keys (as string constants) and values are popped from the stack. The resulting object
+    /// is then pushed back on the stack.
+    CreateObject,
+
+    /// Pushes an empty parameter onto the parameter stack. An empty parameter will be a missing optional parameter
+    /// that will be passed to the next function called.
+    EmptyParameter,
+
+    /// Moves a value from the stack onto the parameter stack in preparation for calling a function.
+    MoveParameter,
+
+    /// Moves a static parameter indicated by the ensuing primitive onto the parameter stack. A static parameter
+    /// will have been created by the `compile_argument` function of the `Function` that is about to be called
+    /// at compile time. (Used, for example, to precompile and store regexes at compile time.)
+    MoveStaticParameter,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Instruction {
+    /// An OpCode is an instruction to the machine to perform some operation.
+    OpCode(OpCode),
+
+    /// Primitives can represent several different things:
+    ///
+    /// Index into constants.
+    /// The amount to jump during a Jump instruction.
+    /// Index into targets for path get and set operations.
+    /// Number of fields whilst building up arrays and objects.
+    /// Index into the list of functions during a call.
+    /// Index into statics for static parameters.
+    Primitive(usize),
+}
```

Comment:
(_Edit:_ After reading the rest of the code, I see that some opcodes (notably `Call`) fetch multiple primitives, which would then cause the size of all variants to grow to four words, invalidating the math below. I still think it is worth exploring, particularly if such opcodes that need more than one value can be broken down into multiple opcodes, but it's not an obvious slam dunk as such.)

I realize this is late in the review cycle, but I haven't seen any commentary about this issue, so I'm leaving this note for posterity or future consideration.

If I am reading the comments in the `OpCode` enum above correctly, each opcode accesses either zero or one primitive, and operates on the stack for the remainder of the data. If so, this is strictly a pessimism of both the memory layout and safety of the instruction stream as compared to storing the primitive (parameter) in the opcode variant directly:
1. Memory layout: opcodes that use zero primitives occupy two words here and four words with one primitive, while storing the parameter with the opcode will take a fixed two words per opcode. Of course, this calculation changes if there is ever an opcode that needs more than one parameter, but this has so far been managed by the use of a evaulation stack for the rest of the data.
2. Safety: When executing the instruction stream, the VM will have to fetch an instruction and effectively `unwrap` the opcode within, and then fetch an instruction and `unwrap` the primitive. With the opcode variant storage, the parameters are part of the variant and require no extra type checking. Now, given that we control the code generation, we can also control that those unwraps never break, but it would be preferable to avoid them in the first place IMO.

I expect this has an impact on the performance of this representation, but that would best be borne out by benchmarking.

Reply:
@fuchsnj had some good ideas early on about how we can optimize this using variable length bytecodes. Essentially it would require us to store the instructions in a `Vec<u8>`, each item would be interpreted as either an instruction or a primitive (which could be more than a single byte.) This would be less safe, but could be a substantial optimization. We decided to stick with maximum safety for the first version, and then look into such optimizations in due course.